README.md
dalek-browser-chrome
=====================

> DalekJS browser plugin for Google Chrome

[![Build Status](https://travis-ci.org/dalekjs/dalek-browser-chrome.png)](https://travis-ci.org/dalekjs/dalek-browser-chrome)
[![Build Status](https://drone.io/github.com/dalekjs/dalek-browser-chrome/status.png)](https://drone.io/github.com/dalekjs/dalek-browser-chrome/latest)
[![Dependency Status](https://david-dm.org/dalekjs/dalek-browser-chrome.png)](https://david-dm.org/dalekjs/dalek-browser-chrome)
[![devDependency Status](https://david-dm.org/dalekjs/dalek-browser-chrome/dev-status.png)](https://david-dm.org/dalekjs/dalek-browser-chrome#info=devDependencies)
[![NPM version](https://badge.fury.io/js/dalek-browser-chrome.png)](http://badge.fury.io/js/dalek-browser-chrome)
[![Coverage](http://dalekjs.com/package/dalek-browser-chrome/master/coverage/coverage.png)](http://dalekjs.com/package/dalek-browser-chrome/master/coverage/index.html)
[![unstable](https://rawgithub.com/hughsk/stability-badges/master/dist/unstable.svg)](http://github.com/hughsk/stability-badges)
[![Stories in Ready](https://badge.waffle.io/dalekjs/dalek-browser-chrome.png?label=ready)](https://waffle.io/dalekjs/dalek-browser-chrome)
[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/dalekjs/dalek-browser-chrome/trend.png)](https://bitdeli.com/free "Bitdeli Badge")
[![Built with Grunt](https://cdn.gruntjs.com/builtwith.png)](http://gruntjs.com/)

[![NPM](https://nodei.co/npm/dalek-browser-chrome.png)](https://nodei.co/npm/dalek-browser-chrome/)
[![NPM](https://nodei.co/npm-dl/dalek-browser-chrome.png)](https://nodei.co/npm/dalek-browser-chrome/)

## Ressources

[API Docs](http://dalekjs.com/package/dalek-browser-chrome/master/api/index.html) -
[Trello](https://trello.com/b/xhw6Jv7A/dalek-browser-chrome) -
[Code coverage](http://dalekjs.com/package/dalek-browser-chrome/master/coverage/index.html) -
[Code complexity](http://dalekjs.com/package/dalek-browser-chrome/master/complexity/index.html) -
[Contributing](https://github.com/dalekjs/dalek-browser-chrome/blob/master/CONTRIBUTING.md) -
[User Docs](http://dalekjs.com/docs/chrome.html) -
[Homepage](http://dalekjs.com) -
[Twitter](http://twitter.com/dalekjs)

## Docs

This module is a browser plugin for [DalekJS](//github.com/dalekjs/dalek).
It provides all a WebDriverServer & browser launcher for Google Chrome.

The browser plugin can be installed with the following command:

```
$ npm install dalek-browser-chrome --save-dev
```

You can use the browser plugin by adding a config option to the your [Dalekfile](/pages/config.html)

```
"browser": ["chrome"]
```

Or you can tell Dalek that it should test in this browser via the command line:

```
$ dalek mytest.js -b chrome
```

The Webdriver Server tries to open Port 9002 by default,
if this port is blocked, it tries to use a port between 9003 & 9092
You can specifiy a different port from within your [Dalekfile](/pages/config.html) like so:

```
"browsers": [{
  "chrome": {
    "port": 5555
  }
}]
```

It is also possible to specify a range of ports:

```
"browsers": [{
  "chrome": {
    "portRange": [6100, 6120]
  }
}]
```

If you would like to test Chrome Canary oder Chromium releases, you can simply apply a snd. argument,
which defines the browser type:

```
$ dalek mytest.js -b chrome:canary
```

for canary, and if you would like to use chromium, just append `:chromium`:

```
$ dalek mytest.js -b chrome:chromium
```

This will only work if you installed your browser in the default locations,
if the browsers binary is located in a non default location, you are able to specify
its location in your [Dalekfile](/pages/config.html):

```javascript
"browsers": [{
  "chrome": {
    "binary": "/Applications/Custom\ Located\ Chrome.app/Contents/MacOS/Google\ Chrome"
  }
}]
```

This also works for the canary & chromium builds

```javascript
"browsers": [{
  "chrome": {
    "binary": "/Applications/Custom\ Located\ Chrome.app/Contents/MacOS/Google\ Chrome"
  }
}]
```

```
$ dalek mytest.js -b chrome
```
### Launch Chrome with flags enabled

Set `chromeOptions` to start a custom instance of Chrome with flags / command lines switches enabled
for various experimental features. [All command line switches for Chrome](http://peter.sh/experiments/chromium-command-line-switches/).

```javascript
"browsers": [{
  "chrome": {
    "chromeOptions": {
      "args": ["enable-experimental-web-platform-features", "js-flags=--harmony"]
    }
  }
}]
```

## Help Is Just A Click Away

### #dalekjs on FreeNode.net IRC

Join the `#daleksjs` channel on [FreeNode.net](http://freenode.net) to ask questions and get help.

### [Google Group Mailing List](https://groups.google.com/forum/#!forum/dalekjs)

Get announcements for new releases, share your projects and ideas that are
using DalekJS, and join in open-ended discussion that does not fit in
to the Github issues list or StackOverflow Q&A.

**For help with syntax, specific questions on how to implement a feature
using DalekJS, and other Q&A items, use StackOverflow.**

### [StackOverflow](http://stackoverflow.com/questions/tagged/dalekjs)

Ask questions about using DalekJS in specific scenarios, with
specific features. For example, help with syntax, understanding how a feature works and
how to override that feature, browser specific problems and so on.

Questions on StackOverflow often turn in to blog posts or issues.

### [Github Issues](//github.com/dalekjs/dalek-browser-chrome/issues)

Report issues with DalekJS, submit pull requests to fix problems, or to
create summarized and documented feature requests (preferably with pull
requests that implement the feature).

**Please don't ask questions or seek help in the issues list.** There are
other, better channels for seeking assistance, like StackOverflow and the
Google Groups mailing list.

![DalekJS](https://raw.github.com/dalekjs/dalekjs.com/master/img/logo.png)

## Legal FooBar (MIT License)

Copyright (c) 2013 Sebastian Golasch

Distributed under [MIT license](https://github.com/dalekjs/dalek-browser-chrome/blob/master/LICENSE-MIT)


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/dalekjs/dalek-browser-chrome/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



CHANGELOG.md
<a name="0.0.10"></a>
### 0.0.10 (2014-11-28)


<a name="0.0.9"></a>
### 0.0.9 (2014-03-10)


<a name="v0.0.8"></a>
### v0.0.8 (2014-01-22)

<a name="v0.0.7"></a>
### v0.0.7 (2013-11-13)

<a name="v0.0.6"></a>
### v0.0.6 (2013-10-29)


#### Bug Fixes

* **core:** fixed process resolving on linux ([ab9d6591](http://github.com/dalekjs/dalek-browser-chrome/commit/ab9d659106ae00a424f08afdc29af0d3ea746469))

<a name="v0.0.5"></a>
### v0.0.5 (2013-10-17)


#### Bug Fixes

* **gruntfile:** Added missing paramters ([84df0635](http://github.com/dalekjs/dalek-browser-chrome/commit/84df06351cf4a2200bacf33e84af683d337dab86))
* **readme:** Typo in image path * 'master' of github.com:dalekjs/dalek-browser-chrome:   fix  ([9a45d2ea](http://github.com/dalekjs/dalek-browser-chrome/commit/9a45d2eab36689ec43dd83ea2ff4913094bc3679))

<a name="v0.0.4"></a>
### v0.0.4 (2013-09-25)

#### Bug Fixes

* **core:**
  * fixed chrome process cant be killed issue ([4fb9545](https://github.com/dalekjs/dalek-browser-chrome/commit/4fb954598586a656c904b3b2abcb999fca2727ef))

<a name="v0.0.3"></a>
### v0.0.3 (2013-09-23)

#### Bug Fixes

* **core:**
  * fixed Windows loader issue with process list management ([f95eefd](https://github.com/dalekjs/dalek-browser-chrome/commit/f95eefd24268c2fdcf8d80091c793895dc795dc4))

<a name="v0.0.2"></a>
### v0.0.2 (2013-09-23)

#### Features

* **install:**
  * updated to chromedriver2 ([787a140](https://github.com/dalekjs/dalek-browser-chrome/commit/787a140d50c762c482dd8721f4194e8df79945d0))

* **core:**
  * added ability to specify ports & port ranges for the user ([787a140](https://github.com/dalekjs/dalek-browser-chrome/commit/787a140d50c762c482dd8721f4194e8df79945d0))

* **gruntfile:**
  * improved build/release automation tools ([787a140](https://github.com/dalekjs/dalek-browser-chrome/commit/787a140d50c762c482dd8721f4194e8df79945d0))

#### Typos

* **readme:**
  * typos in documentation ([b166b9e](https://github.com/dalekjs/dalek-browser-chrome/commit/d17ecda2486a3d1565097dfe1a9e0147c2c17711))

#### Bug Fixes

* **install:**
  * downgraded Chromedriver downloader to not use broken version 0.8 ([b166b9e](https://github.com/dalekjs/dalek-browser-chrome/commit/b166b9ecf65f2be0ff79ec4a5f708a7f2e0bd450))

* **core:**
  * proper shutdown of chromedriver ([60b5fd4](https://github.com/dalekjs/dalek-browser-chrome/commit/60b5fd47dad069d062510c782a5db01fed6122c8))

<a name="v0.0.1"></a>
### v0.0.1 (2013-08-22)

#### Bug Fixes

* **install:**
  * updated install script to latest chromedriver ([b7b6bd7](https://github.com/dalekjs/dalek-browser-chrome/commit/b7b6bd7efde1750f5d7e617566448372b6a2afc9))
  * fixed install script for windows ([0f65f03](https://github.com/dalekjs/dalek-browser-chrome/commit/0f65f0335facdf9f8039c94ca2d4448b3a7ef6f7))

#### Features

* **gruntfile:**
  * added build/release automation tools ([9b8585f](https://github.com/dalekjs/dalek-browser-chrome/commit/9b8585fd54c6f4ce8f06f24e2c97385094e56bb7))


test/index_TEST.js
'use strict';

var expect = require('chai').expect;
var ChromeDriver = require('../index');

describe('dalek-browser-chrome', function() {

  it('should get default webdriver port', function(){
    expect(ChromeDriver.port).to.equal(9002);
  });

});


Gruntfile.js
module.exports = function (grunt) {
  'use strict';

  // check task runtime
  require('time-grunt')(grunt);

  // load generic configs
  var configs = require('dalek-build-tools');

  // project config
  grunt.initConfig({

    // load module meta data
    pkg: grunt.file.readJSON('package.json'),

    // define a src set of files for other tasks
    src: {
      lint: ['Gruntfile.js', 'index.js', 'install.js', 'lib/**/*.js', 'test/*.js'],
      complexity: ['index.js', 'install.js', 'lib/**/*.js'],
      test: ['test/*.js'],
      src: ['index.js', 'install.js', 'lib/**/*.js']
    },

    // clean automatically generated helper files & docs
    clean: configs.clean,

    // speed up build by defining concurrent tasks
    concurrent: configs.concurrent,

    // linting
    jshint: configs.jshint,

    // testing
    mochaTest: configs.mocha,

    // code metrics
    complexity: configs.complexity,
    plato: configs.plato(grunt.file.readJSON('.jshintrc')),

    // api docs
    yuidoc: configs.yuidocs(),

    // user docs
    documantix: {
      options: {
        header: 'dalekjs/dalekjs.com/master/assets/header.html',
        footer: 'dalekjs/dalekjs.com/master/assets/footer.html',
        target: 'report/docs',
        vars: {
          title: 'DalekJS - Documentation - Browser - Google Chrome',
          desc: 'DalekJS - Documentation - Browser - Google Chrome',
          docs: true
        }
      },
      src: ['index.js']
    },

    // add current timestamp to the html document
    includereplace: {
      dist: {
        options: {
          globals: {
            timestamp: '<%= grunt.template.today("dddd, mmmm dS, yyyy, h:MM:ss TT") %>'
          },
        },
        src: 'report/docs/*.html',
        dest: './'
      }
    },

    // up version, tag & commit
    bump: configs.bump({
      pushTo: 'git@github.com:dalekjs/dalek-browser-chrome.git',
      files: ['package.json', 'CONTRIBUTORS.md', 'CHANGELOG.md']
    }),

    // generate contributors file
    contributors: configs.contributors,

    // compress artifacts
    compress: configs.compress,

    // prepare files for grunt-plato to
    // avoid error messages (weird issue...)
    preparePlato: {
      options: {
        folders: [
          'coverage',
          'report',
          'report/coverage',
          'report/complexity',
          'report/complexity/files',
          'report/complexity/files/index_js',
          'report/complexity/files/install_js',
          'report/complexity/files/lib_chromedriver_js'
        ],
        files: [
          'report.history.json',
          'files/index_js/report.history.json',
          'files/install_js/report.history.json',
          'files/lib_chromedriver_js/report.history.json'
        ]
      }
    },

    // prepare files & folders for coverage
    prepareCoverage: {
      options: {
        folders: ['coverage', 'report', 'report/coverage'],
        pattern: '[require("fs").realpathSync(__dirname + "/../index.js")]'
      }
    },

    // list requires that need to be changed
    // for generating a canary build
    'release-canary': {
      options: {
        files: ['index.js']
      }
    },

    // archives the docs if a new version appears
    archive: {
      options: {
        file: 'chrome.html'
      }
    }

  });

  // load 3rd party tasks
  require('load-grunt-tasks')(grunt);
  grunt.loadTasks('./node_modules/dalek-build-tools/tasks');

  // define runner tasks
  grunt.registerTask('lint', 'jshint');
  
  // split test & docs for speed
  grunt.registerTask('test', ['clean:coverage', 'prepareCoverage', 'concurrent:test', 'generateCoverageBadge']);
  grunt.registerTask('docs', ['clean:reportZip', 'clean:report', 'preparePlato', 'documantix', 'includereplace', 'concurrent:docs', 'compress']);
  
  // release tasks
  grunt.registerTask('releasePatch', ['test', 'bump-only:patch', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMinor', ['test', 'bump-only:minor', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMajor', ['test', 'bump-only:major', 'contributors', 'changelog', 'bump-commit']);
  
  // clean, test, generate docs (the CI task)
  grunt.registerTask('all', ['clean', 'test', 'docs']);
  
};


index.js
/*!
 *
 * Copyright (c) 2013 Sebastian Golasch
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 */

'use strict';

// ext. libs
var Q = require('q');
var fs = require('fs');
var cp = require('child_process');
var portscanner = require('portscanner');

// int. libs
var chromedriver = require('./lib/chromedriver');

/**
 * This module is a browser plugin for [DalekJS](//github.com/dalekjs/dalek).
 * It provides all a WebDriverServer & browser launcher for Google Chrome.
 *
 * The browser plugin can be installed with the following command:
 *
 * ```bash
 * $ npm install dalek-browser-chrome --save-dev
 * ```
 *
 * You can use the browser plugin by adding a config option to the your [Dalekfile](/pages/config.html)
 *
 * ```javascript
 * "browser": ["chrome"]
 * ```
 *
 * Or you can tell Dalek that it should test in this browser via the command line:
 *
 * ```bash
 * $ dalek mytest.js -b chrome
 * ```
 *
 * The Webdriver Server tries to open Port 9002 by default,
 * if this port is blocked, it tries to use a port between 9003 & 9092
 * You can specifiy a different port from within your [Dalekfile](/pages/config.html) like so:
 *
 * ```javascript
 * "browsers": [{
 *   "chrome": {
 *     "port": 5555 
 *   }
 * }]
 * ```
 *
 * It is also possible to specify a range of ports:
 *
 * ```javascript
 * "browsers": [{
 *   "chrome": {
 *     "portRange": [6100, 6120] 
 *   }
 * }]
 * ```
 *
 * If you would like to test Chrome Canary oder Chromium releases, you can simply apply a snd. argument,
 * which defines the browser type:
 *
 * ```bash
 * $ dalek mytest.js -b chrome:canary
 * ```
 *
 * for canary, and if you would like to use chromium, just append `:chromium`:
 *
 * ```bash
 * $ dalek mytest.js -b chrome:chromium
 * ```
 *
 * This will only work if you installed your browser in the default locations,
 * if the browsers binary is located in a non default location, you are able to specify
 * its location in your [Dalekfile](/pages/config.html):
 *
 * ```javascript
 * "browsers": [{
 *   "chrome": {
 *     "binary": "/Applications/Custom Located Chrome.app/MacOS/Contents/Chrome" 
 *   }
 * }]
 * ```
 *
 * This also works for the canary &amp; chromium builds
 *
 * ```javascript
 * "browsers": [{
 *   "chrome": {
 *     "binary": "/Applications/Custom Located Chrome.app/MacOS/Contents/Chrome" 
 *   }
 * }]
 * ```
 *
 * @module DalekJS
 * @class ChromeDriver
 * @namespace Browser
 * @part Chrome
 * @api
 */

var ChromeDriver = {

  /**
   * Verbose version of the browser name
   *
   * @property longName
   * @type string
   * @default Google Chrome
   */

  longName: 'Google Chrome',

  /**
   * Default port of the ChromeWebDriverServer
   * The port may change, cause the port conflict resolution
   * tool might pick another one, if the default one is blocked
   *
   * @property port
   * @type integer
   * @default 9002
   */

  port: 9002,

  /**
   * Default maximum port of the ChromeWebDriverServer
   * The port is the highest port in the range that can be allocated
   * by the ChromeWebDriverServer
   *
   * @property maxPort
   * @type integer
   * @default 9092
   */

  maxPort: 9092,

  /**
   * Default host of the ChromeWebDriverServer
   * The host may be overridden with
   * a user configured value
   *
   * @property host
   * @type string
   * @default localhost
   */

  host: 'localhost',

  /**
   * Default desired capabilities that should be
   * transferred when the browser session gets requested
   *
   * @property desiredCapabilities
   * @type object
   */

  desiredCapabilities: {
    browserName: 'chrome',
    chromeOptions: {}
  },

  /**
   * Driver defaults, what should the driver be able to access.
   *
   * @property driverDefaults
   * @type object
   */

  driverDefaults: {
    viewport: true,
    status: true,
    sessionInfo: true
  },

  /**
   * Root path of the ChromeWebDriverServer
   *
   * @property path
   * @type string
   * @default /wd/hub
   */

  path: '/wd/hub',

  /**
   * Child process instance of the Chrome browser
   *
   * @property spawned
   * @type null|Object
   * @default null
   */

  spawned: null,

  /**
   * Chrome processes that are running on startup,
   * and therefor shouldn`t be closed
   *
   * @property openProcesses
   * @type array
   * @default []   
   */

  openProcesses: [],

  /**
   * Name of the process (platform dependent)
   * that represents the browser itself
   *
   * @property processName
   * @type string
   * @default chrome.exe / Chrome   
   */

  processName: (process.platform === 'win32' ? 'chrome.exe' : 'Chrome'),

  /**
   * Different browser types (Canary / Chromium) that can be controlled
   * via the Chromedriver
   *
   * @property browserTypes
   * @type object
   */

  browserTypes: {

    /**
     * Chrome Canary
     *
     * @property canary
     * @type object
     */

    canary: {
      name: 'Chrome Canary',
      linux: 'google-chrome-canary',
      darwin: '/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary',
      win32: process.env.LOCALAPPDATA + '\\Google\\Chrome SxS\\Application\\chrome.exe'
    },

    /**
     * Chromium
     *
     * @property chromium
     * @type object
     */

    chromium: {
      name: 'Chromium',
      process: (process.platform === 'win32' ? 'chromium.exe' : 'Chromium'),
      linux: 'chromium-browser',
      darwin: '/Applications/Chromium.app/Contents/MacOS/Chromium',
      win32: process.env.LOCALAPPDATA + '\\Google\\Chrome SxS\\Application\\chrome.exe'
    }
  },

  /**
   * Resolves the driver port
   *
   * @method getPort
   * @return {integer} port WebDriver server port
   */

  getPort: function () {
    return this.port;
  },

  /**
   * Resolves the maximum range for the driver port
   *
   * @method getMaxPort
   * @return {integer} port Max WebDriver server port range
   */

  getMaxPort: function () {
    return this.maxPort;
  },

  /**
   * Returns the driver host
   *
   * @method getHost
   * @return {string} host WebDriver server hostname
   */

  getHost: function () {
    return this.host;
  },

  /**
   * Launches the ChromeWebDriverServer
   * (which implicitly launches Chrome itself)
   * and checks for an available port
   *
   * @method launch
   * @param {object} configuration Browser configuration
   * @param {EventEmitter2} events EventEmitter (Reporter Emitter instance)
   * @param {Dalek.Internal.Config} config Dalek configuration class
   * @return {object} promise Browser promise
   */

  launch: function (configuration, events, config) {
    var deferred = Q.defer();

    // store injected configuration/log event handlers
    this.reporterEvents = events;
    this.configuration = configuration;
    this.config = config;

    // check for a user set port
    var browsers = this.config.get('browsers');
    if (browsers && Array.isArray(browsers)) {
      browsers.forEach(this._checkUserDefinedPorts.bind(this));
    }

    if (configuration) {
      if (configuration.chromeOptions) {
        this.desiredCapabilities.chromeOptions = configuration.chromeOptions;
      }

      // check for a user defined binary
      if (configuration.binary) {
        var binaryExists = this._checkUserDefinedBinary(configuration.binary);
        if (binaryExists) {
          // check for new verbose & process name
          this.longName = this._modifyVerboseBrowserName(configuration);
          this.processName = this._fetchProcessName(configuration);
        }
      }
    }

    // check if the current port is in use, if so, scan for free ports
    portscanner.findAPortNotInUse(this.getPort(), this.getMaxPort(), this.getHost(), this._checkPorts.bind(this, deferred));
    return deferred.promise;
  },

  /**
   * Kills the ChromeWebDriverServer
   * & Chrome browser processes
   *
   * @method kill
   * @chainable
   */

  kill: function () {
    this._processes(process.platform, this._checkProcesses.bind(this));
    return this;
  },

  /**
   * Modifies the verbose browser name
   *
   * @method _modifyVerboseBrowserName
   * @param {object} configuration User configuration
   * @return {string} Verbose browser name
   * @private
   */

  _modifyVerboseBrowserName: function (configuration) {
    if (configuration.type && this.browserTypes[configuration.type]) {
      return this.browserTypes[configuration.type].name + ' (' + this.longName + ')';
    }

    return this.longName;
  },

  /**
   * Change the process name for browser instances like Canary &amp; Chromium
   *
   * @method _fetchProcessName
   * @param {object} configuration User configuration
   * @return {string} Verbose browser name
   * @private
   */

  _fetchProcessName: function (configuration) {
    // check if the process name must be overridden (to shut down the browser)
    if (this.browserTypes[configuration.type] && this.browserTypes[configuration.type].process) {
      return this.browserTypes[configuration.type].process;
    }

    return this.processName;
  },

  /**
   * Process user defined ports
   *
   * @method _checkUserDefinedPorts
   * @param {object} browser Browser configuration
   * @chainable
   * @private
   */

  _checkUserDefinedPorts: function (browser) {
    // check for a single defined port
    if (browser.chrome && browser.chrome.port) {
      this.port = parseInt(browser.chrome.port, 10);
      this.maxPort = this.port + 90;
      this.reporterEvents.emit('report:log:system', 'dalek-browser-chrome: Switching to user defined port: ' + this.port);
    }

    // check for a port range
    if (browser.chrome && browser.chrome.portRange && browser.chrome.portRange.length === 2) {
      this.port = parseInt(browser.chrome.portRange[0], 10);
      this.maxPort = parseInt(browser.chrome.portRange[1], 10);
      this.reporterEvents.emit('report:log:system', 'dalek-browser-chrome: Switching to user defined port(s): ' + this.port + ' -> ' + this.maxPort);
    }

    return this;
  },

  /**
   * Checks if the binary exists,
   * when set manually by the user
   *
   * @method _checkUserDefinedBinary
   * @param {string} binary Path to the browser binary
   * @return {bool} Binary exists
   * @private
   */

  _checkUserDefinedBinary: function (binary) {
    // check if we need to replace the users home directory
    if (process.platform === 'darwin' && binary.trim()[0] === '~') {
      binary = binary.replace('~', process.env.HOME);
    }
    
    // check if the binary exists
    if (!fs.existsSync(binary)) {
      this.reporterEvents.emit('error', 'dalek-driver-chrome: Binary not found: ' + binary);
      process.exit(127);
      return false;
    }

    // add the binary to the desired capabilities
    this.desiredCapabilities.chromeOptions.binary = binary;

    return true;
  },

  /**
   * Checks if the def. port is blocked & if we need to switch to another port
   * Kicks off the process manager (for closing the opened browsers after the run has been finished)
   * Also starts the chromedriver instance 
   *
   * @method _checkPorts
   * @param {object} deferred Promise
   * @param {null|object} error Error object
   * @param {integer} port Found open port
   * @private
   * @chainable
   */

  _checkPorts: function(deferred, error, port) {
    // check if the port was blocked & if we need to switch to another port
    if (this.port !== port) {
      this.reporterEvents.emit('report:log:system', 'dalek-browser-chrome: Switching to port: ' + port);
      this.port = port;
    }

    // get the currently running processes & invoke the chromedriver afterwards
    this._processes(process.platform, this._startChromedriver.bind(this, deferred));
    return this;
  },

  /**
   * Spawns an instance of Chromedriver
   * 
   * @method _startChromedriver
   * @param {object} deferred Promise
   * @param {null|object} error Error object
   * @param {string} result List of open chrome processes BEFORE the test browser has been launched
   * @private
   * @chainable
   */

  _startChromedriver: function (deferred, err, result) {
    var args = ['--port=' + this.getPort(), '--url-base=' + this.path];
    this.spawned = cp.spawn(chromedriver.path, args);
    this.openProcesses = result;
    this.spawned.stdout.on('data', this._catchDriverLogs.bind(this, deferred));
    return this;
  },

  /**
   * Watches the chromedriver console output to capture the starting message
   * 
   * @method _catchDriverLogs
   * @param {object} deferred Promise
   * @param {buffer} data Chromedriver console output
   * @private
   * @chainable
   */

  _catchDriverLogs: function (deferred, data) {
    var dataStr = data + '';
    var timeout = null;

    // timeout to catch if chromedriver couldnt be launched
    if (dataStr.search('DVFreeThread') === -1) {
      timeout = setTimeout(function () {
        deferred.reject();
        this.reporterEvents.emit('error', 'Chromedriver: ' + dataStr.trim());
        this.reporterEvents.emit('error', 'dalek-driver-chrome: Could not launch Chromedriver');
        process.exit(127);
      }.bind(this), 2000);
    }

    // look for the success message
    if (dataStr.search('Starting ChromeDriver') !== -1) {
      this.reporterEvents.emit('report:log:system', 'dalek-browser-chrome: Started ChromeDriver');
      if (timeout !== null) {
        clearTimeout(timeout);
      }
      deferred.resolve();
    }

    return this;
  },

  /**
   * Remove the chromedriver log that is written to the current working directory
   * 
   * @method _unlinkChromedriverLog
   * @param {bool} retry Delete has been tried min 1 time before
   * @private
   * @chainable
   */

  _unlinkChromedriverLog: function (retry) {
    var logfile = process.cwd() + '/chromedriver.log';
    try {
      if (fs.existsSync(logfile)) {
        fs.unlinkSync(logfile);
      }
    } catch (e) {
      if (!retry) {
        setTimeout(this._unlinkChromedriverLog.bind(this, true), 1000);
      }
    }
    
    return this;
  },

  /**
   * Tracks running browser processes for chrome on mac & linux
   *
   * @method _processes
   * @param {string} platform Current OS
   * @param {function} fn Callback
   * @chainable
   * @private
   */

  _processes: function (platform, fn) {
    if (platform === 'win32') {
      this._processesWin(fn);
      return this;
    }
    
    this._processesNix(fn);
    return this;
  },

  /**
   * Kills all associated processes
   * 
   * @method _checkProcesses
   * @param {object|null} err Error object or null
   * @param {array} result List of running processes
   * @chainable
   * @private
   */

  _checkProcesses: function (err, result) {
    // log that the driver shutdown process has been initiated
    this.reporterEvents.emit('report:log:system', 'dalek-browser-chrome: Shutting down ChromeDriver');
    // kill leftover chrome browser processes
    result.forEach(this[(process.platform === 'win32' ? '_killWindows' : '_killNix')].bind(this));
    // kill chromedriver binary      
    this.spawned.kill('SIGTERM');
    // clean up the file mess the chromedriver leaves us behind
    this._unlinkChromedriverLog();
    return this;
  },

  // UNIX ONLY
  // ---------

  /**
   * Kills a process
   * 
   * @method _killNix
   * @param {integer} processID Process ID
   * @chainable
   * @private
   */

  _killNix: function (processID) {
    var kill = true;
    this.openProcesses.forEach(function (pid) {
      if (pid === processID) {
        kill = false;
      }
    });

    if (kill === true) {
      var killer = cp.spawn;
      killer('kill', [processID]);
    }

    return this;
  },

  /**
   * Lists all chrome processes on *NIX systems
   * 
   * @method _processesNix
   * @param {function} fn Calback
   * @chainable
   * @private
   */

  _processesNix: function (fn) {
    var processName = process.platform === 'darwin' ? this.processName : this.processName.toLowerCase();
    var cmd = ['ps -ax', '|', 'grep ' + processName];
    cp.exec(cmd.join(' '), this._processListNix.bind(this, fn));
    return this;
  },

  /**
   * Deserializes a process list on nix
   * 
   * @method _processListNix
   * @param {function} fn Calback
   * @param {object|null} err Error object
   * @param {string} stdout Output of the process list shell command
   * @chainable
   * @private
   */

  _processListNix: function(fn, err, stdout) {
    var result = [];
    stdout.split('\n').forEach(this._splitProcessListNix.bind(this, result));
    fn(err, result);
    return this;
  },

  /**
   * Reformats the process list output on *NIX systems
   * 
   * @method _splitProcessListNix
   * @param {array} result Reference to the process list
   * @param {string} line Single process in text representation
   * @chainable
   * @private
   */

  _splitProcessListNix: function(result, line) {
    var data = line.split(' ');
    data = data.filter(this._filterProcessItemsNix.bind(this));
    result.push(data[0]);
    return this;
  },

  /**
   * Filters empty process list entries on *NIX
   * 
   * @method _filterProcessItemsNix
   * @param {string} item Item to check
   * @return {string|bool} Item or falsy
   * @private
   */

  _filterProcessItemsNix: function (item) {
    if (item !== '') {
      return item;
    }
    return false;
  },

  // WIN ONLY
  // --------

  /**
   * Lists all running processes (win only)
   *
   * @method _processesWin
   * @param {Function} callback Receives the process object as the only callback argument
   * @chainable
   * @private
   */

  _processesWin: function (callback) {
    cp.exec('tasklist /FO CSV', this._processListWin.bind(this, callback));
    return this;
  },

  /**
   * Deserializes the process list on win
   * 
   * @method _processListWin
   * @param {function} callback Callback to be executed after the list has been transformed
   * @param {object|null} err Error if error, else null
   * @param {string} stdout Output of the process list command
   * @chainable
   * @private
   */

  _processListWin: function (callback, err, stdout) {
    var p = [];
    stdout.split('\r\n').forEach(this._splitProcessListWin.bind(this, p));
    var proc = [];
    var head = null;
    while (p.length > 1) {
      var rec = p.shift();
      rec = rec.replace(/\"\,/gi,'";').replace(/\"|\'/gi,'').split(';');
      if (head === null){
        head = rec;
        for (var i=0;i<head.length;i++){
          head[i] = head[i].replace(/ /gi,'');
        }
      } else {
        var tmp = {};
        for (var j=0;j<rec.length;j++){
          tmp[head[j]] = rec[j].replace(/\"|\'/gi,'');
        }
        proc.push(tmp);
      }
    }

    var result = [];
    proc.forEach(this._filterProcessItemsWin.bind(this, result));
    callback(null, result);
    return this;
  },

  /**
   * Processes (transforms) the list of processes
   * 
   * @method _filterProcessItemsWin
   * @param {array} result Reference to the result process list
   * @param {object} process Single piece of process information
   * @chainable
   * @private
   */

  _filterProcessItemsWin: function (result, process) {
    Object.keys(process).forEach(function (key) {
      if (process[key] === this.processName) {
        result.push(process.PID);
      }
    }.bind(this));
    return this;
  },

  /**
   * Filters empty lines out of the process result
   * 
   * @method _splitProcessListWin
   * @param {array} p Reference to the process list
   * @param {string} line Process item
   * @chainable
   * @private
   */

  _splitProcessListWin: function (p, line) {
    if (line.trim().length !== 0) {
      p.push(line);
    }
    return this;
  },

  /**
   * Kills a process (based on a PID)
   * that was not opened BEFORE Dalek has
   * been started
   *
   * @method _killWindows
   * @param {integer} pid Process id
   * @chainable
   * @private
   */

  _killWindows: function (pid) {
    var kill = true;

    // check if the process was running BEFORE we started dalek
    this.openProcesses.forEach(function (opid) {
      if (opid === pid) {
        kill = false;
      }
    });

    // kill the browser process
    if (kill === true) {
      cp.exec('taskkill /PID ' + pid + ' /f', function(){});
    }

    return this;
  }

};

// expose the module
module.exports = ChromeDriver;
